loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+across Quilt, this system will allow us to remove the special case code for Mixin
+and access wideners and reimplement those as loader plugins instead. This will make
+developing third-party libraries that modify bytecode in unique and interesting
+ways possible.
+
+Loader plugins could also be used to [add mod containers for mods that don't do anything](https://github.com/FabricMC/fabric-loader/issues/175), although a
+better system for that will likely be implemented separately into Loader itself.
+
+A loader plugin could be developed to load mods from instance-specific folders instead of the single shared mods folder.
+
+By breaking parts of the mod loading process into various projects, we can keep
+each repository focused on its own specific mission. This will allow for more
+finely grained separation of responsibility.
+
+
+# Explanation
+
+The addition of Quilt Loader plugins will make Quilt Loader a plugin loader
+first, and mod loader second. Much of the existing mod loading behavior will be
+moved to one or more first-party loader plugins that are installed by default,
+keeping only the bare minimum amount of code required for mod resolving in Loader
+proper. Loader plugins will be capable of effecting the mod loading process by
+adding/removing mods and even adding other plugins. Plugins will also have the
+ability to open/create UI elements which can be used for as simple a thing as
+prompting the user for confirmation, or as complex of one as writing a full-scale
+mod launcher as a loader plugin.
+
+## Application
+Loader plugins are initially added from the classpath, allowing default loader
+plugins for an instance to be added via the profile json. All plugins are then
+applied in cycles. During each cycle, the following things happen:
+- Each plugin is given a chance to add any plugin candidates
+    - Each new plugin added this way is given a chance to add any plugin candidates.
+      Rinse and repeat.
+- Each plugin is given a chance to add any required or tentative mod candidates,
+  rules, or options
+- Quilt Loader will evaluate the sat4j tree
+    - Each plugin will be given a chance to claim unsatisfied rules for resolution
+    - Each plugins resolvers will be run sequentially
+
+If any changes to the mod set have taken place during a given cycle, another cycle
+will be run. Otherwise, the game will either launch or crash, depending on whether
+or not all rules have been satisfied.
+
+
+## API Surface
+
+Below is a tentative plan of what the surface level API classes might look like.
+Everything here is subject (and likely) to change, but helps provide a visual
+idea of how developers might use this feature down the road.
+
+<details>
+    <summary>QuiltPluginContext</summary>
+
+```java
+/**
+* Passed to loader plugins to define what actions they are able to take.
+*/
+public sealed interface QuiltPluginContext permits QuiltPluginContextImpl {
```

Comment:
I assume this is just to show this is not for implementation by users. Since loader does still need to support java 8 of course.

Reply:
It is, though support for Java 8 is another conversation that should probably be had. I don't think bending over backwards to support versions earlier than 1.17 should be a priority for us.

Reply:
for the other issue but a note: supporting versions newer than 8 is possible but not the reverse if loader is compiled for java 16